autofaiss/external/distributed.py

```diff
@@ -0,0 +1,372 @@
+            .appName("Distributed autofaiss")
+            .getOrCreate()
+        )
+    return ss
+
+
+def _batch_loader(batch_size: int, nb_batches: int) -> Iterator[Tuple[int, int, int]]:
+    """Yield [batch id, batch start position, batch end position]"""
+    for batch_id in range(nb_batches):
+        start = batch_size * batch_id
+        end = batch_size * (batch_id + 1)
+        yield batch_id, start, end
+
+
+def _download_one(src_dst_path: Tuple[str, str], fs: fsspec.AbstractFileSystem):
+    src_path, dst_path = src_dst_path
+    fs.get(src_path, dst_path)
+
+
+def _parallel_download_indices_from_remote(
+    fs: fsspec.AbstractFileSystem, indices_file_paths: List[str], dst_folder: str
+):
+    """Download small indices in parallel."""
+    if len(indices_file_paths) == 0:
+        return
+    os.makedirs(dst_folder, exist_ok=True)
+    nb_files = len(indices_file_paths)
+    dst_paths = [os.path.join(dst_folder, os.path.split(p)[-1]) for p in indices_file_paths]
+    src_dest_paths = zip(indices_file_paths, dst_paths)
+    with tqdm(total=nb_files) as pbar:
+        with ThreadPool(min(50, len(indices_file_paths))) as pool:
+            for _ in pool.imap_unordered(partial(_download_one, fs=fs), src_dest_paths):
+                pbar.update(1)
+
+
+def _merge_index(small_indices_folder: str, max_size_on_disk: str = "100GB") -> faiss.Index:
+    """Merge all the indices in `small_indices_folder` into single one return the merged index."""
+    fs = _get_file_system(small_indices_folder)
+    small_indices_files = fs.ls(small_indices_folder, detail=False)
+    if len(small_indices_files) == 0:
+        raise ValueError(f"No small index is saved in {small_indices_folder}")
+
+    def _get_index_from_file(filepath: str) -> faiss.Index:
+        with open(filepath, "rb") as f:
+            index_bytes = f.read()
+        return _get_index_from_bytes(index_bytes)
+
+    def _merge_from_local(merged: Optional[faiss.Index] = None) -> faiss.Index:
+        local_file_paths = [
+            os.path.join(local_indices_folder, filename) for filename in os.listdir(local_indices_folder)
+        ]
+        if merged is None:
+            merged = _get_index_from_file(local_file_paths[0])
+            start_index = 1
+        else:
+            start_index = 0
+
+        for rest_index_file in tqdm(local_file_paths[start_index:]):
+            index = _get_index_from_file(rest_index_file)
+            faiss.merge_into(merged, index, shift_ids=True)
+        return merged
+
+    # estimate index size by taking the first index
+    first_index_file = small_indices_files[0]
+    first_index_size = fs.size(first_index_file)
+    max_sizes_in_bytes = cast_memory_to_bytes(max_size_on_disk)
+    nb_files_each_time = math.floor(max_sizes_in_bytes / first_index_size)
+    merged_index = None
+    n = len(small_indices_files)
+    nb_iterations = max(math.ceil(n / nb_files_each_time), 1)
+    local_indices_folder = "/tmp/distributed_autofaiss_indices"
```

Comment:
Better use https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory for this
Otherwise there's a risk that this folder would already exist